.coveragerc

```diff
@@ -1,7 +1,25 @@
 [run]
+source_pkgs =
+	setuptools
+	pkg_resources
 omit =
 	# leading `*/` for pytest-dev/pytest-cov#456
-	*/.tox/*
+	*/_vendor/*
+	*/_distutils/*
+	*/tests/*
+	*/test_*
+
+[paths]
+setuptools =
+	setuptools/
+	*/site-packages/setuptools/
+	*/pip-install-*/setuptools_*/setuptools/
+	*/tmp*/setuptools
+pkg_resources =
+	pkg_resources/
+	*/site-packages/pkg_resources/
+	*/pip-install-*/setuptools_*/pkg_resources/
+	*/tmp*/pkg_resources
```

Comment:
The combination of the project layout and the kinds of tests we run is rather unique (2 top-level packages, in-tree tests and flat-layout, tests using pip-run), which require a bespoke path configuration...

For most of packages using src-layout, the following configuration is enough:
```ini
[paths]
source =
	src/
	*/site-packages/
```
The equivalent for flat-layout is:
```ini
[paths]
source =
	../  # seems to be resolved against the `.coveragerc` file path, not the current folder
	*/site-packages/
```

... but `coverage` does not allow specifying wildcards as the last segment in `[paths]`, therefore we need an entry for each `setuptools` and `pkg_resources`.
`*/pip-install-*` paths seem to be created by `pip-run` and `*/tmp*` paths are created when installing into directories created with the `tempfile` module.

Reply:
This makes me sad. I've noticed recently (3-6 months ago) in my other projects that they too have started reporting coverage (or errors measuring coverage) on files outside the source tree. Making this change here implies adding a similar but unique change to every other project.

What I'd really like to see is some automation in pytest-cov or coverage to automatically detect the files in the local repo and use those files to determine where to measure coverage.

I've filed jaraco/skeleton#56 to track the broader issue. Can we start by investigating the issue in a simpler project like jaraco/tempora and solve other meta issues that affect other projects in the skeleton before applying solutions that are setuptools-specific and add more maintenance burden to this project?

Reply:
Thank you very much for the comment guys.

I spent quite a lot of time the past week trying to investigate on how to make this configuration as generic as possible, experimenting with several projects, but unfortunately I could not find that "one configuration to rule them all".

There seem to be a lot of factors that can influence the report, for example:

- flat vs src-layout
- In-package tests vs separated `tests` folder
- Usage of `develop=True` in `tox.ini`
- Usage of `isolated_build=True` in `tox.ini`
- Usage of `changedir` in `tox.ini`
- Usage of `--installpkg` flag with `tox`
- `coverage` vs `pytest-cov`

Some things I have noticed:

- When using a flat layout and `develop=True`, omitting the `[paths]` seems to work OK. However that does not work well with src-layout...
- When I was experimenting with `tempora`, for some reason, using `--cov` alone seems to cause coverage warnings, while using `--cov tempora` results in a cleaner output, **regardless** of the presence of `[run] source = tempora` in `.coveragerc`.

So far the best approach I did manage to find (less noisy) is to use `[run] source = ...`, `[paths]` and still pass in the command line `--cov=PACKAGE_NAME` instead of just `--cov`. That seems to work for a broad combination of the factors I described previously, but it does need the name of the package under test to be hardcoded in some configuration files...